materialize-databricks/sqlgen.go

```diff
@@ -48,64 +26,36 @@ var databricksDialect = func() sql.Dialect {
-
-	type customDDLFieldConfig struct {
-		DDL string `json:"DDL,omitempty"`
-	}
-
-	var customDDLMapper = func(mapper sql.TypeMapper) sql.FieldConfigMapper[customDDLFieldConfig] {
-		return sql.FieldConfigMapper[customDDLFieldConfig]{
-			Delegate: mapper,
-			Map: func(m *sql.MappedType, config customDDLFieldConfig) error {
-				if config.DDL != "" {
-					m.DDL = config.DDL
-				}
-
-				return nil
-			},
-		}
-	}
+	var jsonMapper = sql.MapStatic("STRING", sql.UsingConverter(sql.ToJsonString))
 
 	// https://docs.databricks.com/en/sql/language-manual/sql-ref-datatypes.html
-	var mapper sql.TypeMapper = sql.ProjectionTypeMapper{
-		sql.ARRAY:    jsonMapper,
-		sql.BINARY:   sql.NewStaticMapper("BINARY"),
-		sql.BOOLEAN:  sql.NewStaticMapper("BOOLEAN"),
-		sql.INTEGER:  customDDLMapper(sql.NewStaticMapper("BIGINT")),
-		sql.NUMBER:   sql.NewStaticMapper("DOUBLE"),
-		sql.OBJECT:   jsonMapper,
-		sql.MULTIPLE: jsonMapper,
-		sql.STRING: sql.StringTypeMapper{
-			Fallback: sql.NewStaticMapper("STRING"),
-			WithFormat: map[string]sql.TypeMapper{
-				"integer": sql.PrimaryKeyMapper{
-					PrimaryKey: sql.NewStaticMapper("STRING"),
-					Delegate:   sql.NewStaticMapper("BIGINT", sql.WithElementConverter(sql.StdStrToInt())),
+	mapper := sql.NewDDLMapper(
+		sql.FlatTypeMappings{
+			sql.ARRAY:   jsonMapper,
+			sql.BINARY:  sql.MapStatic("BINARY"),
+			sql.BOOLEAN: sql.MapStatic("BOOLEAN"),
+			sql.INTEGER: sql.MapSignedInt64(
+				sql.MapStatic("LONG", sql.UsingConverter(sql.CheckedInt64)),
+				sql.MapStatic("NUMERIC(38,0)", sql.AlsoCompatibleWith("decimal")),
+			),
+			sql.NUMBER:   sql.MapStatic("DOUBLE"),
+			sql.OBJECT:   jsonMapper,
+			sql.MULTIPLE: jsonMapper,
+			sql.STRING_INTEGER: sql.MapStringMaxLen(
+				sql.MapStatic("NUMERIC(38,0)", sql.AlsoCompatibleWith("decimal"), sql.UsingConverter(sql.StrToInt)),
+				sql.MapStatic("STRING", sql.UsingConverter(sql.ToStr)),
+				38,
+			),
+			sql.STRING_NUMBER: sql.MapStatic("DOUBLE", sql.UsingConverter(sql.StrToFloat("NaN", "Inf", "-Inf"))),
+			sql.STRING: sql.MapString(sql.StringMappings{
+				Fallback: sql.MapStatic("STRING"),
+				WithFormat: map[string]sql.MapProjectionFn{
+					"date":      sql.MapStatic("DATE"),
+					"date-time": sql.MapStatic("TIMESTAMP"),
 				},
-				"number": sql.PrimaryKeyMapper{
-					PrimaryKey: sql.NewStaticMapper("STRING"),
-					Delegate:   sql.NewStaticMapper("DOUBLE", sql.WithElementConverter(sql.StdStrToFloat("NaN", "Inf", "-Inf"))),
-				},
-				"date":      sql.NewStaticMapper("DATE"),
-				"date-time": sql.NewStaticMapper("TIMESTAMP"),
-			},
+			}),
 		},
-	}
-
-	mapper = sql.NullableMapper{
-		NotNullText: "NOT NULL",
-		Delegate:    mapper,
-	}
-
-	columnValidator := sql.NewColumnValidator(
-		sql.ColValidation{Types: []string{"string"}, Validate: stringCompatible},
-		sql.ColValidation{Types: []string{"boolean"}, Validate: sql.BooleanCompatible},
-		sql.ColValidation{Types: []string{"long", "decimal"}, Validate: sql.IntegerCompatible},
```

Comment:
Note that "decimal" was added here in https://github.com/estuary/connectors/pull/1570 to support a specific use case. The new custom DDL capability handles this by allowing fields with custom DDL to pass validation with any column type, why is the only practical thing we can really do.